### PR TITLE
Add option to ban former passwords

### PR DESCRIPTION
**Pull request by mfrister**
OpenProject Issue: https://www.openproject.org/issues/778

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -690,6 +683,21 @@ def candidates_for_project_allowance project
     @registered_allowance_evaluators.map{ |f| f.project_granting_candidates(project) }.flatten.uniq
   end
 
+  def current_password
+    self.passwords.first
```
Comment (ulferts) [inline app/models/user.rb]: You did not specify an order here. Neither did you specify a default scope in UserPassword. So you will receive an arbitrary password (based on the db).

**Comment by mfrister** [comment]
We still need to decide what to do with the Apache Subversion authentication module.

See https://www.openproject.org/issues/778

File: app/models/user.rb [hunk 2]
```diff
@@ -692,6 +690,17 @@ def candidates_for_project_allowance project
     @registered_allowance_evaluators.map{ |f| f.project_granting_candidates(project) }.flatten.uniq
   end
 
+  def former_passwords_include?(password)
+    return false if Setting[:password_count_former_banned].to_i == 0
+    passwords.any?{ |f| f.same_as_plain_password?(password) }
```
Comment (cratz) [inline app/models/user.rb]: If we 'limit' the passwords to 'Setting[:password_count_former_banned]' than we use the correct number of former passwords for the first password change after lowering  'password_count_former_banned' .

File: app/models/user_password.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  # Hash a plaintext password with a given salt
+  # The hashed password has following form: SHA1(salt + SHA1(password))
+  def self.hash_with_salt(plain_password, salt)
+    # We should really use a standard key-derivation function like bcrypt here
+    hash_password("#{salt}#{hash_password plain_password}")
+  end
+
+  # constant-time comparison algorithm to prevent timing attacks
+  def self.secure_compare(a, b)
```
Comment (cratz) [inline app/models/user_password.rb]: This method returns a bool so the name should end with '?'.

**Comment by mfrister** [comment]
I made the proposed changes and squashed everything to the three commits, let's see what Travis says.

**Comment by mfrister** [comment]
Added copyright notices and changelog. I should start reading the [code review guidelines](https://www.openproject.org/projects/openproject/wiki/Code_Review_Guidelines) I wrote.

**Pull request merged**